apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+    Map.put(current_state, :pending_effects, [
+      %{effect | delay: effect.delay - 1} | List.delete(current_state.pending_effects, effect)
+    ])
   end
 
-  defp choose_targets(_unit, %{targeting_strategy: nil}, _state), do: nil
+  # Check if the unit can attack this turn.
+  # For now, attacking capability is only affected by whether the unit is currently casting a skill.
+  # Later on, things like stuns will be handled here.
+  defp can_attack(unit, initial_step_state) do
+    # Check the unit is not casting anything right now
+    not Enum.any?(initial_step_state.skills_being_cast, &(&1.caster_id == unit.id))
+  end
+
+  # Check if the unit can cast their ultimate skill this step.
+  defp can_cast_ultimate_skill(unit), do: unit.energy >= @ultimate_energy_cost
+
+  # Check if the unit can cast their basic skill this step.
+  defp can_cast_basic_skill(unit), do: unit.basic_skill.remaining_cooldown <= 0
+
+  # Called when a skill being cast reaches effect_trigger 0.
+  # "Queues" the effect to be processed when its delay reaches 0.
+  defp trigger_skill_effects(skill, current_state, initial_step_state) do
+    caster = current_state.units[skill.caster_id]
 
+    # We store the caster's state in the effect in case the unit dies before the effect's delay ends
+    effects_with_caster =
+      Enum.map(skill.effects, fn effect ->
+        effect |> Map.put(:caster, caster) |> Map.put(:targets, choose_targets(caster, effect, initial_step_state))
+      end)
+
+    Map.put(current_state, :pending_effects, effects_with_caster ++ current_state.pending_effects)
+  end
+
+  # Choose the targets for an effect with "random" as the strategy. Returns the target ids.
+  # The `== target_allies` works as a negation operation when `target_allies` is `false`, and does nothing when `true`.
   defp choose_targets(
-         %{team: team},
+         %{team: team} = _caster,
          %{
-           targeting_strategy: "random",
-           targets_allies: targets_allies,
-           amount_of_targets: amount
+           target_count: count,
+           target_strategy: "random",
+           target_allies: target_allies
          } = _effect,
          state
        ),
        do:
-         state
-         |> Enum.filter(fn {_id, unit} -> unit.team == team == targets_allies end)
-         |> Enum.take_random(amount)
+         state.units
+         |> Enum.filter(fn {_id, unit} -> unit.team == team == target_allies end)
+         |> Enum.take_random(count)
          |> Enum.map(fn {id, _unit} -> id end)
 
-  defp apply_effect(%{type: "instant"} = effect, caster, target) do
-    new_value = calculate_value(effect, caster, target)
-    Map.put(target, String.to_atom(effect.stat_affected), new_value)
+  # Apply an effect to its target *if it hits*.
+  # Hit chance is affected by the ChanceToApply component and could be expanded later on.
+  # Returns the new state of the target.
+  defp maybe_apply_effect(effect, target, caster, current_step_number) do
```

Comment:
I think `maybe_apply_effect` is only working a wrapper of an if statement for `effect_hits` and `apply_effect`, maybe we could delete `apply_effect` and do it in the same `maybe_apply_effect` method

Reply:
I think it's clearer to separate between behaviors. When we start to add more mechanics like dodge or accuracy, we'll end up with functions that are way too big. Separating them now is a good way to reduce function overloads both currently and in the long run

Reply:
hmmm maybe we can solve this with pattern match checking if the effect should be applied in the call?

```
maybe_apply_effect(effect, current_state.units[id], current_state.step_number, effect_hits?(effect))

defp maybe_apply_effect(_effect, target, _current_step_number, false = _should_apply?), do: target
defp maybe_apply_effect(%{caster: caster} = effect, target, current_step_number, should_apply?), do: destroy target
```

Reply:
I tried getting rid of the maybe_apply_effect function and ran into the same "function body nested too deep" credo error.

If you find calling the `effect_hits?` function from within `maybe_apply_effect`'s params clearer than an `if` clause, say the word and we'll go for it. I really do not see the advantage in readibility but it may be because of me getting used to the code I wrote.

Reply:
yeah i think it's better because we can get rid of the `apply_effect` method and just do what we need to do without calling anything else. I'm open to debate tho

Reply:
We'll just agree to disagree then 😅

I'll implement your function with the boolean as parameter!